provider/eventstore/eventstore.go

```diff
@@ -0,0 +1,636 @@
+	version              int64
+	globalSequenceNumber uint64
+	savedStreams         map[string]struct{}
+}
+
+// Option defines functional option parameters for eventStore.
+type Option func(*eventStore)
+
+// WithClock is a function option to inject a clock.Clock
+func WithClock(clock clock.Clock) Option {
+	return func(store *eventStore) {
+		store.clock = clock
+	}
+}
+
+// WithUUIDGenerator is a functional option to inject a shortuuid.Generator.
+func WithUUIDGenerator(uuidGenerator shortuuid.Generator) Option {
+	return func(store *eventStore) {
+		store.uuidGenerator = uuidGenerator
+	}
+}
+
+// WithSerializer is a functional option to inject a rangedb.RecordSerializer.
+func WithSerializer(serializer rangedb.RecordSerializer) Option {
+	return func(store *eventStore) {
+		store.serializer = serializer
+	}
+}
+
+// New constructs an eventStore. Experimental: Use at your own risk.
+func New(ipAddr, username, password string, options ...Option) *eventStore {
+	s := &eventStore{
+		clock:        systemclock.New(),
+		serializer:   jsonrecordserializer.New(),
+		broadcaster:  broadcast.New(broadcastRecordBuffSize, broadcast.DefaultTimeout),
+		savedStreams: make(map[string]struct{}),
+		ipAddr:       ipAddr,
+		username:     username,
+		password:     password,
+	}
+
+	for _, option := range options {
+		option(s)
+	}
+
+	return s
+}
+
+func (s *eventStore) NewClient() (*esclient.Client, error) {
+	config, err := esclient.ParseConnectionString(fmt.Sprintf("esdb://%s:%s@%s", s.username, s.password, s.ipAddr))
+	if err != nil {
+		return nil, fmt.Errorf("unexpected configuration error: %s", err.Error())
+	}
+
+	config.DisableTLS = true
+	config.SkipCertificateVerification = true
+	client, err := esclient.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client: %s", err.Error())
+	}
+
+	return client, nil
+}
+
+func (s *eventStore) streamName(name string) string {
+	s.sync.RLock()
+	defer s.sync.RUnlock()
+	return fmt.Sprintf("%d-%s", s.version, name)
+}
+
+func (s *eventStore) SetVersion(version int64) {
+	s.sync.Lock()
+	s.globalSequenceNumber = 0
+	s.version = version
+	s.sync.Unlock()
+}
+
+func (s *eventStore) Bind(events ...rangedb.Event) {
+	s.serializer.Bind(events...)
+}
+
+func (s *eventStore) Events(ctx context.Context, globalSequenceNumber uint64) rangedb.RecordIterator {
+	resultRecords := make(chan rangedb.ResultRecord)
+
+	go func() {
+		defer close(resultRecords)
+
+		client, err := s.NewClient()
+		if err != nil {
+			resultRecords <- rangedb.ResultRecord{
+				Record: nil,
+				Err:    err,
+			}
+			return
+		}
+		defer client.Close()
+
+		readStream, err := client.ReadAllEvents(ctx, direction.Forwards, stream_position.Start{}, ^uint64(0), true)
+		if err != nil {
+			if strings.Contains(err.Error(), rpcErrContextCanceled) {
+				resultRecords <- rangedb.ResultRecord{
+					Record: nil,
+					Err:    context.Canceled,
+				}
+				return
+			}
+
+			if strings.Contains(err.Error(), streamNotFound) {
```

Comment:
This is bad, we should talk to ES DB folks ... checking error message name stuff 😨